### ISSUE TITLE
Cannot shrink vPool when ALBA policy has not been satisfied

### ISSUE THREAD

**Issue by kvanhijf** [openvstorage/framework#1349]
Encountered this issue by executing the following:
* 4 node setup with vPool extended over all nodes
* Node 1, 2 and 3 each had 1 ASD initialized and claimed and i used default policy (2, 2, 3, 4) to create a vPool
* Removed node 3 entirely along with the ASD --> Policy was not satisfied anymore
* Shrinking the vPool now failed with below error:
```
Jan 17 14:19:38 OVS-1-193-151 celery[5918]: 2017-01-17 14:19:38 77900 +0100 - OVS-1-193-151 - 6001/139967615330048 - lib/storagerouter - 59 - INFO - Remove Storage Driver - Guid fa35fdb4-b11f-406a-bd44-25fd42349eb9 - Virtual Disk ab6e3b1
5-a8c0-4634-a7d9-fda129d531c3 vdisk1 - Ensuring MDS safety
Jan 17 14:19:38 OVS-1-193-151 celery[5918]: 2017-01-17 14:19:38 78000 +0100 - OVS-1-193-151 - 6001/139967615330048 - lib/mds - 60 - DEBUG - MDS safety: vDisk ab6e3b15-a8c0-4634-a7d9-fda129d531c3: Start checkup for virtual disk vdisk1
Jan 17 14:19:38 OVS-1-193-151 celery[5918]: 2017-01-17 14:19:38 80200 +0100 - OVS-1-193-151 - 6001/139967615330048 - lib/mds - 61 - DEBUG - MDS safety: vDisk ab6e3b15-a8c0-4634-a7d9-fda129d531c3: Reconfiguration required. Reasons:
Jan 17 14:19:38 OVS-1-193-151 celery[5918]: 2017-01-17 14:19:38 80200 +0100 - OVS-1-193-151 - 6001/139967615330048 - lib/mds - 62 - DEBUG - MDS safety: vDisk ab6e3b15-a8c0-4634-a7d9-fda129d531c3:    * Slave (10.100.193.154:26300) cannot
be used anymore
Jan 17 14:19:39 OVS-1-193-151 celery[5918]: 2017-01-17 14:19:39 11300 +0100 - OVS-1-193-151 - 6001/139967615330048 - lib/mds - 63 - ERROR - MDS safety: vDisk ab6e3b15-a8c0-4634-a7d9-fda129d531c3: Failed to update the metadata backend con
figuration
Jan 17 14:19:39 OVS-1-193-151 celery[5918]: Traceback (most recent call last):
Jan 17 14:19:39 OVS-1-193-151 celery[5918]:   File "/opt/OpenvStorage/ovs/lib/mdsservice.py", line 624, in ensure_safety
Jan 17 14:19:39 OVS-1-193-151 celery[5918]:     req_timeout_secs=5)
Jan 17 14:19:39 OVS-1-193-151 celery[5918]: RuntimeError: got fault response updateMetaDataBackendConfig
Jan 17 14:19:39 OVS-1-193-151 celery[5918]: 2017-01-17 14:19:39 11300 +0100 - OVS-1-193-151 - 6001/139967615330048 - lib/storagerouter - 64 - ERROR - Remove Storage Driver - Guid fa35fdb4-b11f-406a-bd44-25fd42349eb9 - Virtual Disk ab6e3b
15-a8c0-4634-a7d9-fda129d531c3 vdisk1 - Ensuring MDS safety failed
Jan 17 14:19:39 OVS-1-193-151 celery[5918]: Traceback (most recent call last):
Jan 17 14:19:39 OVS-1-193-151 celery[5918]:   File "/opt/OpenvStorage/ovs/lib/storagerouter.py", line 985, in remove_storagedriver
Jan 17 14:19:39 OVS-1-193-151 celery[5918]:     excluded_storagerouters=[storage_router] + storage_routers_offline)
Jan 17 14:19:39 OVS-1-193-151 celery[5918]:   File "/opt/OpenvStorage/ovs/lib/mdsservice.py", line 627, in ensure_safety
Jan 17 14:19:39 OVS-1-193-151 celery[5918]:     raise Exception('MDS configuration for volume {0} with guid {1} could not be changed'.format(vdisk.name, vdisk.guid))
Jan 17 14:19:39 OVS-1-193-151 celery[5918]: Exception: MDS configuration for volume vdisk1 with guid ab6e3b15-a8c0-4634-a7d9-fda129d531c3 could not be changed
```
* This resulted in another error further on
```
Jan 17 14:19:45 OVS-1-193-151 celery[5918]: 2017-01-17 14:19:45 71600 +0100 - OVS-1-193-151 - 6001/139967615330048 - lib/storagerouter - 79 - ERROR - Remove Storage Driver - Guid fa35fdb4-b11f-406a-bd44-25fd42349eb9 - Removing MDS service failed
Jan 17 14:19:45 OVS-1-193-151 celery[5918]: Traceback (most recent call last):
Jan 17 14:19:45 OVS-1-193-151 celery[5918]:   File "/opt/OpenvStorage/ovs/lib/storagerouter.py", line 1089, in remove_storagedriver
Jan 17 14:19:45 OVS-1-193-151 celery[5918]:     allow_offline=not storage_router_online)
Jan 17 14:19:45 OVS-1-193-151 celery[5918]:   File "/opt/OpenvStorage/ovs/lib/mdsservice.py", line 179, in remove_mds_service
Jan 17 14:19:45 OVS-1-193-151 celery[5918]:     raise RuntimeError('Cannot remove MDSService that is still serving disks')
Jan 17 14:19:45 OVS-1-193-151 celery[5918]: RuntimeError: Cannot remove MDSService that is still serving disks
```
* And in volumedriver log
```
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 107851 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/XMLRPCTimingWrapper - 0000000000003fe9 - info - execute: Arguments for updateMetaDataBa
ckendConfig are {[metadata_backend_config:FgAAAAAAAABzZXJpYWxpemF0aW9uOjphcmNoaXZlDAAECAQIAQAAAAAAAAAAAQAmAAAAAAAA
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: AHZvbHVtZWRyaXZlcjo6TURTTWV0YURhdGFCYWNrZW5kQ29uZmlnAQMAAAAAAAAAAAAAAAAC
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: AAAAAAAAAAEAAAAAAQAAAA4AAAAAAAAAMTAuMTAwLjE5My4xNTG8Zg4AAAAAAAAAMTAuMTAw
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: LjE5My4xNTK8ZgEAAAAUAAAAAAAAAA==,volume_id:1126a678-d182-44c8-b406-61936b72457c,vrouter_cluster_id:ad81f89c-cc4f-40ff-8ac8-0997c65f3c08]}
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 107903 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/MDSMetaDataStore - 0000000000003fea - info - set_config: 1126a678-d182-44c8-b406-61936b
72457c: new config:
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 107916 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/MDSMetaDataStore - 0000000000003feb - info - set_config:         apply scrub results to
 slaves: ApplyRelocationsToSlaves::T
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 107925 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/MDSMetaDataStore - 0000000000003fec - info - set_config:         mds://10.100.193.151:2
6300
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 107934 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/MDSMetaDataStore - 0000000000003fed - info - set_config:         mds://10.100.193.152:2
6300
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 107943 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/MDSMetaDataStore - 0000000000003fee - info - set_config: 1126a678-d182-44c8-b406-61936b
72457c: mds://10.100.193.151:26300 is already in use, no need to failover
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 108230 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/AlbaProxyClient - 0000000000003fef - info - logMessage: TCPProxy_client(10.100.193.151,
 26205)
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 108341 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/AlbaProxyClient - 0000000000003ff0 - info - logMessage: RoraProxy_client(...)
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 108418 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/BackendConnectionInterfaceLogger - 0000000000003ff1 - info - Logger: Entering write 112
6a678-d182-44c8-b406-61936b72457c volume_configuration
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 111018 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/AlbaConnection - 0000000000003ff2 - error - convert_exceptions_: write object: caught A
lba proxy exception: Proxy_protocol.Protocol.Error.NoSatisfiablePolicy
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 111055 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/BackendConnectionInterfaceLogger - 0000000000003ff3 - error - ~Logger: Exiting write fo
r 1126a678-d182-44c8-b406-61936b72457c volume_configuration with exception
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 111068 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/BackendConnectionInterfaceLogger - 0000000000003ff4 - error - ~Logger: Exiting write fo
r 1126a678-d182-44c8-b406-61936b72457c
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 111080 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/BackendInterface - 0000000000003ff5 - error - do_wrap_: Problem with connection 0x7f1da
80020e0: Proxy_protocol.Protocol.Error.NoSatisfiablePolicy
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 111135 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/AlbaProxyClient - 0000000000003ff6 - info - logMessage: TCPProxy_client(10.100.193.151,
 26205)
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 111223 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/AlbaProxyClient - 0000000000003ff7 - info - logMessage: RoraProxy_client(...)
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 111291 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/BackendInterface - 0000000000003ff8 - warning - do_wrap_: Retrying with new connection
(retry: 1, sleep before retry: 0 milliseconds)
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 111317 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/BackendConnectionInterfaceLogger - 0000000000003ff9 - info - Logger: Entering write 112
6a678-d182-44c8-b406-61936b72457c volume_configuration
Jan 17 14:19:39 OVS-1-193-151 volumedriver_fs.sh[28135]: 2017-01-17 14:19:39 112409 +0100 - OVS-1-193-151 - 28135/0x00007f1dd7ffb700 - volumedriverfs/AlbaConnection - 0000000000003ffa - error - convert_exceptions_: write object: caught A
lba proxy exception: Proxy_protocol.Protocol.Error.NoSatisfiablePolicy
```

Conclusion, when such error is thrown we should catch this and act accordingly

**Comment by wimpers** [comment]
Before removing the node, we should check that the policy can still be satisfied for writing new objects and is not leading to dataloss for existing objects. once you are in a status with unsatisfiable policy,  we can't ensure that things still work.

**Comment by kvanhijf** [comment]
@wimpers : Removing the node? I was shrinking my vPool cluster. The amount of ASDs remained the same at that moment. I'd expect to be able to shrink a vPool cluster regardless of safety? We allow the creation and extension of a vPool too when the policy has not been met, so why not shrinking?

**Comment by wimpers** [comment]
After discussion with @kvanhijf it was clear that he first removed a node rendering the cluster doomed as too many disks are removed.

Updated https://github.com/openvstorage/framework-alba-plugin/issues/282 to also execute a check when removing a node.

**Comment by wimpers** [comment]
The removal of the node should have been blocked. The vPool errors are a logical consequence.

**Comment by kvanhijf** [comment]
@wimpers : What do we do when a node is dead / offline?  Obviously we cannot block the removal in such case?

**Comment by wimpers** [comment]
First add a new node with ASDs.

**Comment by kvanhijf** [comment]
Issue can actually still occur when a node is offline and removal is being postponed. In such case shrinking the vPool will still render the vPool unusable

Edit: Removing an online node, should indeed check whether the policies can still be satisfied

**Comment by wimpers** [comment]
@kvanhijf  what is still open for this ticket? Do we still need to implement the check whether the policies can still be satisfied in case or removing a node?

**Comment by kvanhijf** [comment]
Yes
Shrinking a vPool with unsatisfiable policies should still be tested / verified
Removing an online node which would result in unsatisfiable policies should be tested

PS: Perhaps integration tests should be written too?
Cause i'm not really confident about our code currently with regards removing online/offline nodes in a cluster, since the fwk doesn't test this frequently

**Comment by wimpers** [comment]
Let's use this ticket to add a check in the shrink vPool to first check if  policy can be satisfied, if not don't allow to shrink.  The check should be added in the wizard after you click the delete button (we don't want to  execute the check on the vPool details page for every storage router)

We will create a different ticket for block of the removal of a node in case it would lead to an unsatisfiable policy. https://github.com/openvstorage/framework/issues/1885

**Comment by sjovdnbos** [comment]
Fixed in #1937 
Package information: openvstorage-core_2.11.1-dev.1516612813.390cbf4-1_amd64.deb

**Comment by JeffreyDevloo** [comment]
Part of the https://github.com/openvstorage/framework/releases/tag/2.11.1 release
